File: miband-sdk/src/main/java/com/khmelenko/lab/miband/ObserverWrapper.java
package com.khmelenko.lab.miband;

import io.reactivex.Emitter;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

/**
 * Wraps {@link Observer} with {@link Emitter}
 *
 * @author Dmytro Khmelenko (d.khmelenko@gmail.com)
 */
class ObserverWrapper<T> implements Observer<T> {

    private final Emitter<T> mEmitter;

    public ObserverWrapper(Emitter<T> emitter) {
        mEmitter = emitter;
    }

    @Override
    public void onSubscribe(Disposable d) {
        // do nothing
    }

    @Override
    public void onNext(T value) {
        mEmitter.onNext(value);
    }

    @Override
    public void onError(Throwable e) {
        mEmitter.onError(e);
    }

    @Override
    public void onComplete() {
        mEmitter.onComplete();
    }
}

File: miband-sdk-kotlin/src/main/java/com/khmelenko/lab/miband/ObserverWrapper.kt
package com.khmelenko.lab.miband

import io.reactivex.Emitter
import io.reactivex.Observer
import io.reactivex.disposables.Disposable

/**
 * Wraps [Observer] with [Emitter]

 * @author Dmytro Khmelenko
 */
internal class ObserverWrapper<T>(private val emitter: Emitter<T>) : Observer<T> {

    override fun onSubscribe(d: Disposable) {
        // do nothing
    }

    override fun onNext(value: T) {
        emitter.onNext(value)
    }

    override fun onError(e: Throwable) {
        emitter.onError(e)
    }

    override fun onComplete() {
        emitter.onComplete()
    }
}
